native/src/secret.rs

```diff
@@ -27,18 +27,16 @@ pub fn neon_secret_box(mut cx: FunctionContext) -> JsResult<JsValue> {
   })?;
 
   let key_bytes = cx.borrow(&js_key, |bytes| bytes.as_slice::<u8>());
-  let key_array: [u8; 32] = key_bytes[0..32]
-    .try_into()
-    .or_else(|_| cx.throw_error("expected `secretbox` key to be at least 32 bytes"))?;
-  let key = secretbox::Key(key_array);
-
-  let nonce_array: [u8; 24] = key_bytes[0..24]
-    .try_into()
-    .or_else(|_| cx.throw_error("expected `secretbox` key to be at least 24 bytes"))?;
-  let nonce = secretbox::Nonce(nonce_array);
+  if key_bytes.len() < 32 {
+    cx.throw_error("expected `secretbox` key to be at least 32 bytes")?;
+  }
+  let key = Key::from_slice(&key_bytes[0..32]).unwrap(); // infallible
+  let nonce = Nonce::from_slice(&key_bytes[0..24]).unwrap(); // infallible
 
-  let cyphertext = secretbox::seal(plaintext.as_slice(), &nonce, &key);
-  let buffer = cx.compute_scoped(|mut cx2| utils::bytes_to_buffer(&mut cx2, &cyphertext))?;
+  let hmac = key.seal(&mut plaintext, &nonce);
+  // `plaintext` now contains the cyphertext. "Attached" format begins with the hmac:
+  plaintext.splice(0..0, hmac.0.iter().cloned());
```

Comment:
I don't understand this part, but I'm just going to trust you and the tests. :+1:

Reply:
secretbox::Key::seal encrypts the plaintext in place, to avoid allocating a new buffer. The `splice` part is an ugly way (and the only way I'm aware of) to insert the hmac bytes into the front of the vec. `0..0` is an empty Range, and splice "replaces" the items in that range with the items in the iterator created by hmac.0.iter().cloned() (which is an ugly way of getting an iterator where the items are u8, instead of &u8). We could skip all this by allocating a new buffer of length Hmac::SIZE + plaintext.len() and using key.seal_into_attached (or the convenience function that could/should exist that would do that for you), but that would cost us some precious nanoseconds!